src/arena.c

```diff
@@ -1297,32 +1345,43 @@ arena_tcache_fill_small(tsdn_t *tsdn, arena_t *arena, tcache_t *tcache,
 	bin = &arena->bins[binind];
 	malloc_mutex_lock(tsdn, &bin->lock);
 	for (i = 0, nfill = (tcache_bin_info[binind].ncached_max >>
-	    tcache->lg_fill_div[binind]); i < nfill; i++) {
+	    tcache->lg_fill_div[binind]); i < nfill; i += cnt) {
 		extent_t *slab;
-		void *ptr;
+		cnt = 1;
 		if ((slab = bin->slabcur) != NULL && extent_nfree_get(slab) >
 		    0) {
-			ptr = arena_slab_reg_alloc(slab, &bin_infos[binind]);
+			unsigned tofill = nfill - i;
+			cnt = tofill < extent_nfree_get(slab) ?
+				tofill : extent_nfree_get(slab);
+			arena_slab_reg_alloc_batch(
+			   slab, &bin_infos[binind], cnt,
+			   tbin->avail - nfill + i);
 		} else {
-			ptr = arena_bin_malloc_hard(tsdn, arena, bin, binind);
-		}
-		if (ptr == NULL) {
+			void *ptr = arena_bin_malloc_hard(tsdn, arena, bin,
```

Comment:
Another small optimization we may try: split `arena_bin_malloc_hard` into `bin_slab_refill` and calling `slab_reg_alloc`. In this case we'd only need the refill part and continue the loop with cnt = 0. OOM handling might be a bit tricky (maybe fall back to bin_malloc_hard if re-fill didn't get a new nonfull slab).

Reply:
I split the function in https://github.com/jemalloc/jemalloc/pull/1338, I'll rebase that one and then test again

Reply:
Ah I see. If you want to land this one as is, that's totally fine too.

Reply:
Right, I'll land this one and test this optimization separately